ADAL/tests/integration/ios/ADBrokerIntegrationTests.m

```diff
@@ -193,6 +193,90 @@ - (void)testBroker_whenSimpleAcquireToken_shouldSucceed
+          @"skip_cache" : @"NO",
+          @"extra_qp" : @"",
+          @"claims" : @"",
+          };
+
+        NSString *expectedUrlString = [NSString stringWithFormat:@"msauth://broker?%@", [expectedParams adURLFormEncode]];
+        NSURL *expectedURL = [NSURL URLWithString:expectedUrlString];
+        XCTAssertTrue([expectedURL matchesURL:url]);
+
+        NSDictionary *responseParams =
+        @{
+          @"authority" : authority,
+          @"resource" : TEST_RESOURCE,
+          @"client_id" : TEST_CLIENT_ID,
+          @"id_token" : [[self adCreateUserInformation:TEST_USER_ID] rawIdToken],
+          @"access_token" : @"i-am-a-access-token",
+          @"refresh_token" : @"i-am-a-refresh-token",
+          @"foci" : @"1",
+          @"expires_in" : @"3600",
+          @"error_code" : @"213", // AD_ERROR_SERVER_PROTECTION_POLICY_REQUIRED
+          @"error_description" : @"AADSTS53005: Application needs to enforce intune protection policies",
+          @"error" : @"unauthorized_client",
+          };
+
+        [ADAuthenticationContext handleBrokerResponse:[ADBrokerIntegrationTests createV2BrokerResponse:responseParams redirectUri:redirectUri]];
+        return YES;
+    }];
+
+    NSArray *metadata = @[ @{ @"preferred_network" : @"login.microsoftonline.com",
+                              @"preferred_cache" : @"login.windows.net",
+                              @"aliases" : @[ @"login.windows.net", @"login.microsoftonline.com"] } ];
+    ADTestURLResponse *validationResponse =
+    [ADTestAuthorityValidationResponse validAuthority:authority
+                                          trustedHost:@"login.windows.net"
+                                         withMetadata:metadata];
+    [ADTestURLSession addResponses:@[validationResponse]];
+
+    ADAuthenticationContext *context = [self getBrokerTestContext:authority];
+    XCTestExpectation *expectation = [self expectationWithDescription:@"acquire token callback"];
+    [context acquireTokenWithResource:TEST_RESOURCE
+                             clientId:TEST_CLIENT_ID
+                          redirectUri:[NSURL URLWithString:redirectUri]
+                      completionBlock:^(ADAuthenticationResult *result)
+     {
+         XCTAssertNotNil(result);
+         XCTAssertEqual(result.status, AD_FAILED);
+         XCTAssertEqual(result.error.code, AD_ERROR_SERVER_PROTECTION_POLICY_REQUIRED);
+
+         [expectation fulfill];
+     }];
+
+    [self waitForExpectations:@[expectation] timeout:1.0];
+
+    // Assert that MAM token was added to cache even though broker returned an error code
+    ADKeychainTokenCache *tokenCache = (ADKeychainTokenCache *)[context tokenCacheStore].dataSource;
+
+    XCTAssertEqualObjects([tokenCache getAT:authority], @"i-am-a-access-token");
+    XCTAssertEqualObjects([tokenCache getMRRT:authority], @"i-am-a-refresh-token");
+    XCTAssertEqualObjects([tokenCache getFRT:authority], @"i-am-a-refresh-token");
+}
```

Comment:
could we add a unit test to test the case where broker returns AD_ERROR_SERVER_PROTECTION_POLICY_REQUIRED but no MAM tokens? making sure we are robust enough to handle that. Nothing should be stored in cache in such a case.